### PR TITLE
Improved command line interface

### PR DESCRIPTION
**Pull request by DecentM**
Thanks for working on this project! I recently had to write a small script that uses AssetRipper, and noticed that most of the options are missing from the CLI. This PR improves the command line interface, and brings it to feature parity with the GUI.

List of changes:
- Adds command line options for the following settings:
  - Bundled assets export mode
  - Ignore StreamingAssets folder
  - Audio export format
  - Image export format
  - Mesh export format
  - Sprite export mode
  - Terrain export mode
  - Text export mode
  - Shader export mode
  - Script export mode
  - Script export level
  - Script language version
- Inverts the `-q` logic and renames it to `-k` (`--keep-open`). It's much more likely that a command line user will expect AssetRipper to exit after it's done, as the printed logs will be available on the user's screen even after it quits, provided that AssetRipper was launched from an already open terminal. In case it's launched through a shortcut that opens its own terminal, the `-k` option is available to get the previous behaviour.
- Fixes compiler warnings, as CommandLine provides nullable types for some of its options. These values are then checked before they're used.
- Makes `-o` a required parameter, it looked like it was meant to be required, as it had no default.
- The `--help` output now has an expanded `.WithNotParsed` handler so it can print enum values. In debug builds, it also prints out the error names that caused the help screen to be printed as well.
- Makes small code style changes so the same style is used throughout `ConsoleApp.cs`.

### CODE REVIEW

**Comment by DecentM** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by DecentM** [comment]
recheck

**Comment by DecentM** [comment]
recheck

**Comment by ds5678** [comment]
I greatly appreciate your willingness to contribute to the project. Unfortunately, I think I have to reject this pr outright. There's several problems:

* The feature disparity was not accidental. These options require much higher maintenance than the previous set of options. This maintenance cost almost always falls to me since most contributors don't stick around. Any pr increasing the number of command line options should make it exceedingly easy for me to maintain that new standard of options. In other words, adding or changing a configuration option should be so easy that it incurs almost no maintenance cost at all. This might be handled best with some code in the GUI source generator project.
* Your changes to the -q and -o options directly affect me in an adverse way. They were designed that way because it's easiest for me. This new behavior would be a constant thorn in my side.
* Your changes deepen the use of the current command line library, which I seek to be rid of. It's not trim compatible. I would prefer a switch to `System.CommandLine`.

Once again, I appreciate your willingness to contribute. If you decide to rewrite this pull request in the manner I outlined above, I'd be happy to reconsider a larger set of command line options.

**Comment by DecentM** [comment]
It looks like a switch from `CommandLine` to `System.CommandLine` lines up well with code generation for the CLI feature as a whole, since `System.CommandLine` uses functions to create its options, while `CommandLine` uses tags. I'll look into the codegen project and try to implement it that way.  

I'm curious, how does inverting the `-q` logic affect you negatively? When I used the command line for the first time, I let AssetRipper hang after it finished for a good 5 minutes as I thought it was still processing things. Normally CLIs exit once done, with sometimes an optional keepalive-like parameter, so it seemed like a no-brainer for me. The actual functionality remains the same, but just with different arguments.  

With `-o`, it looked like I didn't change the actual behaviour by making it required. Was the output directory set to the workdir when it wasn't provided by the user? It'd be trivial to add that as a default, but it wasn't clear from just reading the code.

**Comment by ds5678** [comment]
I just drop a file onto the executable. Inverting the logic would make the window close before I want it to.

Yes, it has a convenient default.

**Comment by DecentM** [comment]
I see, how about this logic then:  

If a file is dropped onto the executable, it will get the canonical path to that file in the first argument. So if the first argument is a path, AssetRipper will assume that it's being run from explorer by dropping a file onto it and automatically enable `-k`. When it's run from the terminal using named arguments, it will expect the user to supply `-k` in order not to quit at the end.  

This flow both enables your use as is from before, but it also behaves in the standard way when called by the user typing into their terminal.

**Comment by ds5678** [comment]
If the argument array has length 1 and the singular argument does not start with `-`, assume the file explorer method of execution. Otherwise, assume command prompt usage.

Does a keep alive option have significant utility? It seems like it could be removed entirely. At the very least, it is undeserving of being a single letter argument.

**Comment by DecentM** [comment]
Good point, I'll remove the `-k` option for now, and we can add it back later if needed, although I can't think of anything else we would need that option for.  

I'll start pushing wip stuff, please don't merge until it's ready.

**Comment by ds5678** [comment]
FYI I see you using reflection. That is not trim compatible.

**Comment by DecentM** [comment]
Interesting, I didn't know that. I was already moving away from reflection and into just refactoring the things we need to generate cli options.

**Comment by DecentM** [comment]
Alright, I think it's ready for review, here's what I ended up doing:  

If the first argument does not start with a dash, `System.CommandLine` will parse it as the root command, which has (or should have anyway) the same functionality as before. The root command has the following options and arguments:
- `0`: Positional argument, input paths
- `-o` `--output` (defaults to `Ripped`, just so it doesn't spew junk into the repo when testing, could change it to `.` easily if needed)
- `-q` `--quit`
- `-v` `--verbose`
- `-w` `--log-file` <log-file>  

If the first argument is `e` or `extract`, `System.CommandLine` will parse it as the `extract` subcommand. It has all options from the root command, except `-q` `--quit`. In addition, `ConsoleApp` will read all currently stored settings from a new dictionary in `CoreConfiguration`, though the `Ripper` instance, and generate `Option`s from them.  

This approach comes with the limitation that all settings must be defined as `enum`s, so I've converted `ScriptContentLevel`, `IgnoreStreamingAssets`, and `ExportDependencies` to be backed by enums stored in `CoreConfiguration`, while keeping the existing API as-is. The upside to this approach, is that all settings automatically get validated and autocompleted CLI options. To later expand this list, all you'll need to do is call `CoreConfiguration(instance).SetSettings<SomeEnumType>(settingValue)` before `ConsoleApp.ParseArgumentsAndRun` is called from the main `Program.cs`.

File: .gitignore [hunk 1]
```diff
@@ -12,4 +12,7 @@ bin
 .vscode
 *.blend1
 launchSettings.json
-AssetRipper.SourceGenerated.xml
\ No newline at end of file
+AssetRipper.SourceGenerated.xml
+
+/Ripped
+/AssetRipper.log
```
Comment (ds5678) [inline .gitignore:18]: I fail to see why these were necessary. They should always be created in a subfolder of `Bins` or `bin`.

File: AssetRipper.Fundamentals/Configuration/CoreConfiguration.cs [hunk 2]
```diff
@@ -3,12 +3,39 @@
...
 using AssetRipper.Core.Parser.Files;
 using AssetRipper.Core.Utils;
+using System.Collections.Generic;
 using System.IO;
 
 namespace AssetRipper.Core.Configuration
 {
 	public class CoreConfiguration
 	{
+		public Dictionary<Type, object> settings = new Dictionary<Type, object>();
```
Comment (ds5678) [inline AssetRipper.Fundamentals/Configuration/CoreConfiguration.cs:13]: This uses boxing, which I oppose almost as strongly as reflection.

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/types/boxing-and-unboxing

File: AssetRipper.Fundamentals/Configuration/CoreConfiguration.cs [hunk 3]
```diff
@@ -86,7 +121,7 @@ public void SetProjectSettings(UnityVersion version, BuildTarget platform, Trans
 		public virtual void ResetToDefaultValues()
 		{
 			ScriptContentLevel = ScriptContentLevel.Level2;
-			StreamingAssetsMode = StreamingAssetsMode.Extract;
+			IgnoreStreamingAssets = false;
```
Comment (ds5678) [inline AssetRipper.Fundamentals/Configuration/CoreConfiguration.cs:124]: Why did you change this?

File: AssetRipper.Fundamentals/Configuration/ExportDependencies.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+		/// <summary>
+		/// Dependencies are exported with the project.
+		/// </summary>
+		Export,
+		/// <summary>
+		/// Dependencies are ignored during export.
+		/// </summary>
+		Ignore,
+	}
+}
```
Comment (ds5678) [inline AssetRipper.Fundamentals/Configuration/ExportDependencies.cs:14]: This is not a setting I want exposed. I purposefully didn't include it in the GUI.

File: AssetRipper.GUI.SourceGenerator/Program.cs [hunk 5]
```diff
@@ -4,13 +4,16 @@ namespace AssetRipper.GUI.SourceGenerator;
...
-		
-		string source = LocalizationSourceGenerator.MakeLocalizationClass(repositoryPath);
+		return Path.Combine(repositoryPath, project);
+	}
 
-		File.WriteAllText(outputPath, source);
+	public static void Main(string[] args)
+	{
+		string localizationPath = Path.Combine(GetPathForProject("AssetRipper.GUI"), "LocalizationManager.g.cs");
+		File.WriteAllText(localizationPath, LocalizationSourceGenerator.MakeLocalizationClass(repositoryPath));
```
Comment (ds5678) [inline AssetRipper.GUI.SourceGenerator/Program.cs:17]: This seems completely unrelated to everything else you were doing.

File: AssetRipper.GUI/ConsoleApp.cs [hunk 6]
```diff
@@ -1,126 +1,179 @@
...
-			public IReadOnlyList<string> FilesToExport { get; set; }
+			Logger.AllowVerbose = verbose;
+			Logger.Add(new ConsoleLogger(false));
 
-			[Option('o', "output", HelpText = "Directory to export to. Will be cleared if already exists.")]
-			public DirectoryInfo OutputDirectory { get; set; }
+			// Do not log to a file if the logging target is null. It should be set from ConsoleOptions.DefaultLogFileName
+			// if the user didn't specify one.
+			if (string.IsNullOrEmpty(logPath))
+				return;
```
Comment (ds5678) [inline AssetRipper.GUI/ConsoleApp.cs:25]: A log file should be created always.

File: AssetRipper.GUI/ConsoleApp.cs [hunk 7]
```diff
@@ -1,126 +1,179 @@
...
+
+#if DEBUG
+			Console.WriteLine("===================================================");
+			Console.WriteLine("      Error during ripping, extraction aborted     ");
+			Console.WriteLine("===================================================");
+			Console.WriteLine(ex.ToString());
+#else
+			Console.WriteLine("Error during ripping, extraction aborted");
+			Console.WriteLine("See this message below for details:");
+			Console.WriteLine(ex.Message);
```
Comment (ds5678) [inline AssetRipper.GUI/ConsoleApp.cs:126]: Errors like this must always be saved to the log.

File: AssetRipper.GUI/ConsoleOptions.cs [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+using AssetRipper.Library;
+
+
+namespace AssetRipper.GUI
+{
+	internal enum ExitCode
+	{
+		ExtractingError = 1,
+		ArgumentError = 2,
+	}
```
Comment (ds5678) [inline AssetRipper.GUI/ConsoleOptions.cs:19]: Separate types go in separate files.

File: AssetRipper.Library/Configuration/LibraryConfiguration.cs [hunk 9]
```diff
@@ -1,73 +1,40 @@
...
-		public SpriteExportMode SpriteExportMode { get; set; }
-		/// <summary>
-		/// How terrain data is exported. Recommended: Native
-		/// </summary>
-		public TerrainExportMode TerrainExportMode { get; set; }
-		/// <summary>
-		/// How are text assets exported?
-		/// </summary>
-		public TextExportMode TextExportMode { get; set; }
-
```
Comment (ds5678) [inline AssetRipper.Library/Configuration/LibraryConfiguration.cs:44]: You are deleting documentation.

File: AssetRipper.Library/Configuration/LibraryConfiguration.cs [hunk 10]
```diff
@@ -1,73 +1,40 @@
...
+			Logger.Info(LogCategory.General, $"{nameof(ImageExportFormat)}: {this.GetSetting<ImageExportFormat>()}");
+			Logger.Info(LogCategory.General, $"{nameof(MeshExportFormat)}: {this.GetSetting<MeshExportFormat>()}");
+			Logger.Info(LogCategory.General, $"{nameof(ScriptExportMode)}: {this.GetSetting<ScriptExportMode>()}");
+			Logger.Info(LogCategory.General, $"{nameof(ScriptLanguageVersion)}: {this.GetSetting<ScriptLanguageVersion>()}");
+			Logger.Info(LogCategory.General, $"{nameof(ShaderExportMode)}: {this.GetSetting<ShaderExportMode>()}");
+			Logger.Info(LogCategory.General, $"{nameof(SpriteExportMode)}: {this.GetSetting<SpriteExportMode>()}");
+			Logger.Info(LogCategory.General, $"{nameof(TerrainExportMode)}: {this.GetSetting<TerrainExportMode>()}");
+			Logger.Info(LogCategory.General, $"{nameof(TextExportMode)}: {this.GetSetting<TextExportMode>()}");
 		}
 	}
```
Comment (ds5678) [inline AssetRipper.Library/Configuration/LibraryConfiguration.cs:39]: `ResetToDefaultValues` and `LogConfigurationValues` are exactly the kind of methods which should be source generated.

File: AssetRipper.Library/Exporters/Audio/AudioClipExporter.cs [hunk 11]
```diff
@@ -14,7 +14,7 @@ namespace AssetRipper.Library.Exporters.Audio
 	public sealed class AudioClipExporter : BinaryAssetExporter
 	{
 		public AudioExportFormat AudioFormat { get; }
-		public AudioClipExporter(LibraryConfiguration configuration) => AudioFormat = configuration.AudioExportFormat;
+		public AudioClipExporter(LibraryConfiguration configuration) => AudioFormat = configuration.GetSetting<AudioExportFormat>();
```
Comment (ds5678) [inline AssetRipper.Library/Exporters/Audio/AudioClipExporter.cs:17]: This method of accessing settings is more cumbersome and verbose.

File: AssetRipper.GUI/ConsoleOptions.cs [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+
+		public static Option<FileInfo> logFileOption
+		{
+			get
+			{
+				if (_logFileOption != null)
+					return _logFileOption;
+
+				Option<FileInfo> option = new Option<FileInfo>(name: "--log-file", description: "File to log to");
+				option.AddAlias("-w");
```
Comment (ds5678) [inline AssetRipper.GUI/ConsoleOptions.cs:323]: Why is `-w` an alias for this?

File: AssetRipper.GUI/ConsoleOptions.cs [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+		ExtractingError = 1,
+		ArgumentError = 2,
+	}
+
+	internal static class ConsoleOptions
+	{
+		public const string DefaultLogFileName = "AssetRipper.log";
+		public const string DefaultOutputPath = "Ripped";
+		public const bool DefaultQuit = false;
+		public const bool DefaultVerbose = false;
```
Comment (ds5678) [inline AssetRipper.GUI/ConsoleOptions.cs:26]: Unless necessary, always make constants private.

File: AssetRipper.GUI/ConsoleOptions.cs [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+
+				Option<bool> option = new Option<bool>(name: "--quit", description: "Quit after ripping");
+				option.AddAlias("-q");
+				option.SetDefaultValue(DefaultQuit);
+
+				_quitOption = option;
+
+				return option;
+			}
+		}
```
Comment (ds5678) [inline AssetRipper.GUI/ConsoleOptions.cs:292]: I thought we agreed on getting rid of this.

File: AssetRipper.GUI/ConsoleOptions.cs [hunk 15]
```diff
@@ -0,0 +1,350 @@
...
+
+				Option<bool> option = new Option<bool>(name: "--verbose", description: "Verbose logging output");
+				option.AddAlias("-v");
+				option.SetDefaultValue(DefaultVerbose);
+
+				_verboseOption = option;
+
+				return option;
+			}
+		}
```
Comment (ds5678) [inline AssetRipper.GUI/ConsoleOptions.cs:311]: Why is there so much code required for a simple boolean option?

File: AssetRipper.GUI/ConsoleOptions.cs [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+							return;
+						}
+					}
+				});
+
+				_inputOption = option;
+
+				return option;
+			}
+		}
```
Comment (ds5678) [inline AssetRipper.GUI/ConsoleOptions.cs:237]: Why is there an `--input` option at all? This was an implicit option before, where anything not directly following an option was considered to be part of the input path list.

File: AssetRipper.GUI/ConsoleOptions.cs [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+		{
+			IEnumerable<string> result = input.Name.Select((character) =>
+			{
+				if (char.IsUpper(character))
+					return $"{separator}{char.ToLower(character)}";
+
+				return character.ToString();
+			});
+
+			return $"{separator}{string.Join("", result)}";
```
Comment (ds5678) [inline AssetRipper.GUI/ConsoleOptions.cs:49]: Use a `StringBuilder` to allocate one string instead of dozens.

File: AssetRipper.GUI/ConsoleApp.cs [hunk 18]
```diff
@@ -1,126 +1,179 @@
...
 		}
 
-		private static void PrepareExportDirectory(string path)
+		public static void ParseArgumentsAndRun(string[] args)
 		{
-			if (Directory.Exists(path))
+			Ripper ripper = new();
+			ripper.Settings.ResetToDefaultValues();
+
+			RootCommand rootCommand = new RootCommand();
```
Comment (ds5678) [inline AssetRipper.GUI/ConsoleApp.cs:135]: This deviates from our discussions. We talked about a check for when `args` had length 1.

File: AssetRipper.GUI/ConsoleApp.cs [hunk 19]
```diff
@@ -1,126 +1,179 @@
...
+			rootCommand.SetHandler((InvocationContext context) => RootCommandHandler(context, ripper));
+			rootCommand.AddArgument(ConsoleOptions.inputArgument);
+			rootCommand.AddOption(ConsoleOptions.outputOption);
+			rootCommand.AddOption(ConsoleOptions.quitOption);
+			rootCommand.AddOption(ConsoleOptions.verboseOption);
+			rootCommand.AddOption(ConsoleOptions.logFileOption);
+
+			Dictionary<string, Option> options = ConsoleOptions.GenerateFromRipper(ripper);
+
+			Command extractCommand = new Command("extract");
```
Comment (ds5678) [inline AssetRipper.GUI/ConsoleApp.cs:145]: I am opposed to any command that doesn't start with `-`.

**Review by ds5678 — CHANGES_REQUESTED** [review]
I've reviewed your changes.

> This approach comes with the limitation that all settings must be defined as enums.

This is a significant downside, especially since you box all the settings.

> I'll look into the codegen project and try to implement it that way.

I assume you encountered difficulties with this in some way. What were they?

**Comment by ds5678** [comment]
Perhaps, it might be easier for you to split your pull request in half:
1. A pull request switching the existing library usage to `System.CommandLine`.
2. A pull request reforming the command line options afterwards.

File: .gitignore (hunk 1 above)
Comment (DecentM) [reply .gitignore:18]: I think it creates them in the workdir, while I was working on it, I nearly committed these paths a few times 😅

File: AssetRipper.Fundamentals/Configuration/CoreConfiguration.cs (hunk 2 above)
Comment (DecentM) [reply AssetRipper.Fundamentals/Configuration/CoreConfiguration.cs:13]: If we're to avoid using code duplication, I see only these two paths, either we use reflection (even if we do codegen, it would have to use reflection to get the member names), or we use a centralised store object to house all settings. I guess we could do some kind of overloading to avoid boxing though, but that feels super messy from a maintenance perspective.

File: AssetRipper.Fundamentals/Configuration/CoreConfiguration.cs (hunk 3 above)
Comment (DecentM) [reply AssetRipper.Fundamentals/Configuration/CoreConfiguration.cs:124]: IgnoreStreamingAssets resolves the enum value in its getter so that we can use GetSettings and SetSettings to store its value (https://github.com/AssetRipper/AssetRipper/pull/513/files/b3ceb5272787dd1b17ca43ce1c68b7690a1ce10b#diff-7581cca434414659a49634ea9b50525e26a8a415750b6e8b7b7e1a0e23e9ad0aR55)

File: AssetRipper.Fundamentals/Configuration/ExportDependencies.cs (hunk 4 above)
Comment (DecentM) [reply AssetRipper.Fundamentals/Configuration/ExportDependencies.cs:14]: Interesting, what's the reason for not exposing that? Seems like useful functionality.

File: AssetRipper.GUI.SourceGenerator/Program.cs (hunk 5 above)
Comment (DecentM) [reply AssetRipper.GUI.SourceGenerator/Program.cs:17]: Oh yeah this was some refactoring I started doing in anticipation of writing some more codegen for the feature that validates that valid enum values are passed. I this this is more future proof anyway, but since I ended up not writing new codegen functions, it's an orphaned change.

File: AssetRipper.GUI/ConsoleApp.cs (hunk 6 above)
Comment (DecentM) [reply AssetRipper.GUI/ConsoleApp.cs:25]: A log file is always created, as the default value is set to "AssetRipper.log". This check is to prevent throwing in case that variable is somehow unset between parsing the arguments and actually accessing its value. It could be due to a bug, or anything else, since System.CommandLine is not marked as stable yet.  

Primarily, it acts as a guard because filePath may technically be an empty string.

File: AssetRipper.GUI/ConsoleOptions.cs (hunk 8 above)
Comment (DecentM) [reply AssetRipper.GUI/ConsoleOptions.cs:19]: Sure, it's such a small declaration that I didn't create a new file for it as it'd make it harder to read.

File: AssetRipper.GUI/ConsoleApp.cs (hunk 7 above)
Comment (DecentM) [reply AssetRipper.GUI/ConsoleApp.cs:126]: @DecentM todo

File: AssetRipper.Library/Configuration/LibraryConfiguration.cs (hunk 9 above)
Comment (DecentM) [reply AssetRipper.Library/Configuration/LibraryConfiguration.cs:44]: @DecentM todo

File: AssetRipper.Library/Configuration/LibraryConfiguration.cs (hunk 10 above)
Comment (DecentM) [reply AssetRipper.Library/Configuration/LibraryConfiguration.cs:39]: Good idea, although since they were like this before, I don't think that's in scope for this pr, since it's about just adding missing options to the cli.

File: AssetRipper.Library/Exporters/Audio/AudioClipExporter.cs (hunk 11 above)
Comment (DecentM) [reply AssetRipper.Library/Exporters/Audio/AudioClipExporter.cs:17]: I think the small amount of extra characters is worth it for the benefits it provides, since more types can be added the the settings object without writing related code to drill the new settings down to wherever they're needed.

File: AssetRipper.GUI/ConsoleOptions.cs (hunk 12 above)
Comment (DecentM) [reply AssetRipper.GUI/ConsoleOptions.cs:323]: Stands for `write`, since `-l` was already taken. Which letter would you suggest?

File: AssetRipper.GUI/ConsoleOptions.cs (hunk 13 above)
Comment (DecentM) [reply AssetRipper.GUI/ConsoleOptions.cs:26]: They're actually used here: https://github.com/AssetRipper/AssetRipper/pull/513/files/b3ceb5272787dd1b17ca43ce1c68b7690a1ce10b#diff-23986fb5dd0effce9afa0388c39d7c44c78cd3d25f8517667c5e355c087e2ae7R47-R50

File: AssetRipper.GUI/ConsoleOptions.cs (hunk 14 above)
Comment (DecentM) [reply AssetRipper.GUI/ConsoleOptions.cs:292]: Yep, I actually removed it first but added it back as it was originally to not break existing scripts. Someone may be using AssetRipper in a script already and upgrading to this would break those use cases. Since this pr is just about adding unavailable parameters, I think it's best to consider removing that in the future.

File: AssetRipper.GUI/ConsoleOptions.cs (hunk 15 above)
Comment (DecentM) [reply AssetRipper.GUI/ConsoleOptions.cs:311]: I know right? According to the docs, that's how you're supposed to do it though. I suppose we could create a new struct to house options and then generate them with a function, but then that becomes something that needs maintenance when the options change.  
https://docs.microsoft.com/en-us/dotnet/standard/commandline/define-commands#define-options

File: AssetRipper.GUI/ConsoleOptions.cs (hunk 16 above)
Comment (DecentM) [reply AssetRipper.GUI/ConsoleOptions.cs:237]: This option is part of the `extract` subcommand, the root command doesn't use it. `--extract` does not use positional arguments since if there is one, the root command will run. In shell scripts, positional arguments are much less readable with no other benefit since all options are passed in using the command line anyway.

File: AssetRipper.GUI/ConsoleOptions.cs (hunk 17 above)
Comment (DecentM) [reply AssetRipper.GUI/ConsoleOptions.cs:49]: @DecentM todo

File: AssetRipper.GUI/ConsoleApp.cs (hunk 18 above)
Comment (DecentM) [reply AssetRipper.GUI/ConsoleApp.cs:135]: This actually does do that, but using System.CommandLine properly, instead of just counting args length, which would be quite hacky since we have System.CommandLine already. If there is no subcommand parsed, the root command is ran, which is exactly that behaviour you described.

https://docs.microsoft.com/en-us/dotnet/standard/commandline/define-commands#define-a-root-command

File: AssetRipper.GUI/ConsoleApp.cs (hunk 19 above)
Comment (DecentM) [reply AssetRipper.GUI/ConsoleApp.cs:145]: This is how System.CommandLine works, you have a root command, and then subcommands, like in dotnet for example you have `dotnet build` and `dotnet restore`, that each have different options.  

https://docs.microsoft.com/en-us/dotnet/standard/commandline/syntax#subcommands

**Comment by DecentM** [comment]
I wrote some replies and added todos for myself at some. I think we're spiraling the scope for this quite a bit, I don't want to get stuck with more changes than minimally needed. I think maybe it would be best to delay this until you can migrate to System.CommandLine upstream, so that you can set it up in the way you'd prefer and then it'd be easier to use that as a baseline for generating options.

**Comment by ds5678** [comment]
Not planned
